jobs/kpi-forecasting/kpi-forecasting/pandas_extras.py

```diff
@@ -0,0 +1,29 @@
+import numpy as np
+import pandas as pd
+
+
+def percentile(p: int = 50, name_format: str = "p{:02.0f}"):
+    """A method to calculate percentiles along dataframe axes via the `pandas.agg` method."""
+
+    def f(x):
+        return x.quantile(p / 100)
+
+    f.__name__ = name_format.format(p)
+    return f
+
+
+def aggregate_to_period(
+    df: pd.DataFrame,
+    period: str,
+    aggregation: callable = np.sum,
+    date_col: str = "submission_date",
+) -> pd.DataFrame:
+    """Floor dates to the correct period and aggregate."""
+    if period.lower() in ["day", "month", "year"]:
+        period = period[0]
+    else:
+        raise ValueError(f"Don't know how to floor dates by {period}.")
```

Comment:
Is ValueError the appropriate Exception type for this error?

Reply:
I think it's okay. The [docs](https://docs.python.org/3/library/exceptions.html#ValueError) say:
> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError).

Reply:
That's fair, I'd maybe make the message a little bit more user friendly and include in it what the expected values are.

Reply:
Done!